src/code/z_elf_message.c

```diff
@@ -64,111 +72,126 @@ u32 ElfMessage_CheckCondition(ElfMessage* msg) {
+
+    // if none of the conditions checked were satisfied, there are none to use
+    if (i == 0) {
         return false;
     }
 
-    temp3 = Rand_ZeroFloat(temp1);
-    for (temp1 = 0; temp1 < temp2; temp1++) {
-        if (sp44[temp1]) {
-            if (temp3 > 0) {
-                temp3--;
+    // choose a random number between 0 and the total amount of conditions met
+    rand = Rand_ZeroFloat(i);
+
+    for (i = 0; i < totalChecked; i++) {
+        // keep decrementing the random number until it reaches 0 then use that hint
+        if (conditions[i]) {
+            if (rand > 0) {
+                rand--;
             } else {
                 return true;
             }
         }
-        *msgp += 1;
+
+        *hintsPtr += 1;
     }
 
     return false;
 }
 
-u16 ElfMessage_GetTextFromMsgs(ElfMessage* msg) {
+u16 Hints_GetTextIdFromScript(HintCmd* hints) {
     while (true) {
-        switch (msg->byte0 & 0xE0) {
-            case (ELF_MSG_TYPE_CHECK << 5):
-                if (ElfMessage_CheckCondition(msg)) {
-                    return msg->byte2 | 0x100;
+        switch (hints->byte0 & 0xE0) {
+            case (HINT_CMD_TYPE_CHECK << 5):
+                if (Hints_CheckCondition(hints)) {
+                    return hints->byte2 | 0x100;
                 }
                 break;
-            case (ELF_MSG_TYPE_UNK_1 << 5):
-                if (func_8006BE88(&msg)) {
-                    return msg->byte2 | 0x100;
+
+            case (HINT_CMD_TYPE_CHAIN << 5):
+                if (Hints_CheckConditionChain(&hints)) {
+                    return hints->byte2 | 0x100;
                 }
                 break;
-            case (ELF_MSG_TYPE_UNK_2 << 5):
-                if (func_8006BF1C(&msg)) {
-                    return msg->byte2 | 0x100;
+
+            case (HINT_CMD_TYPE_RANDOM << 5):
+                if (Hints_CheckRandomCondition(&hints)) {
+                    return hints->byte2 | 0x100;
                 }
                 break;
-            case (ELF_MSG_TYPE_SKIP << 5):
-                if (ElfMessage_CheckCondition(msg)) {
-                    msg += msg->byte2;
-                    msg--;
+
+            case (HINT_CMD_TYPE_SKIP << 5):
+                if (Hints_CheckCondition(hints)) {
+                    hints += hints->byte2; // skip the specified amount
+                    hints--;               // decrement by 1 because it will be incremented again below
                 }
                 break;
-            case (ELF_MSG_TYPE_END << 5):
-                return msg->byte2 | 0x100;
+
+            case (HINT_CMD_TYPE_END << 5):
+                return hints->byte2 | 0x100;
+
             default:
                 LOG_STRING("企画外 条件", "../z_elf_message.c", 281); // "Unplanned conditions"
                 ASSERT(0, "0", "../z_elf_message.c", 282);
         }
-        msg++;
+
+        hints++;
     }
 }
 
-u16 ElfMessage_GetSariaText(PlayState* play) {
+u16 Hints_GetSariaTextId(PlayState* play) {
```

Comment:
Aren't saria hints provided through navi? In a way I would argue this justifies treating the system as "navi hints", which are differentiated between saria and C-Up ones, and I think this is what I was going for with the previous naming.

In general I don't know how I feel about having this system as simply "hints" since there are many other types of hints in the game, including gossip stones, some of the npc dialog, navi hovering certain objects, the indicators in the pause menu map, etc
So even if we don't go with "navi hints" I feel like we should use something more specific

Reply:
elliptic and mzx both separately suggested QuestHint as a system name, which imo is not necessary but if others feel "hints" is too generic then I can do that.

I do feel strongly about keeping navis hints named as navi though. Yes sarias hints are provided through navi, but it is still very clearly hints from saria. I dont want to call the navi quest hints after the name of a button on the controller. And the button on the hud literally says navi on it anyway, they are navi hints imo

Reply:
The text ids referenced by Saria hints are Saria talking, even if it's Navi telling (btw I never knew you weren't directly speaking to Saria)

I think Hints as a system name is enough, there are other hints part of the game design but no such dedicated system. (I would be okay with another name, as you want)

Reply:
One could argue that the targeting + C-Up system is equally well a hint system (and not ab initio just for enemies, but it is an actor hint system at least), although it does not (need to) use scripting. (There's also the ones on the world map screen although I can understand if people don't think they should count.) I think it's more sensible to be specific when this is concerned, is someone looking at the use going to instantly understand what it is without reading the files?

Reply:
yea Im not against quest hint, I just personally dont see it as that necessary, especially when the cost is much larger enum names/function names throughout. But after this conversation I will change it to quest hint

On the topic of navi vs c up though, navi enemy descriptions is another reason to why imo it makes sense to make the seperation within this system as navi and saria. It clearly seperates who is the one giving you the hint, as opposed to the button on the controller which also activates hints from a separate system

Reply:
QuestHint seems good to me yeah, it should help distinguishing the different types of hints